include/exadg/convection_diffusion/driver.cpp

```diff
@@ -143,28 +144,124 @@ Driver<dim, Number>::ale_update() const
   time_int_bdf->ale_update();
 }
 
+template<int dim, typename Number>
+void
+Driver<dim, Number>::mark_cells_coarsening_and_refinement(dealii::Triangulation<dim> & tria,
+                                                          VectorType const & solution) const
+{
+  mark_cells_kelly_error_estimator(tria,
+                                   pde_operator->get_dof_handler(),
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  // Update mapping
+  AssertThrow(ale_mapping.get() == 0,
+              dealii::ExcMessage(
+                "Combination of adaptive mesh refinement and ALE not implemented."));
+
+  std::shared_ptr<dealii::MappingQCache<dim>> mapping_q_cache =
+    std::dynamic_pointer_cast<dealii::MappingQCache<dim>>(mapping);
+  AssertThrow(
+    mapping_q_cache.get() == 0,
+    dealii::ExcMessage(
+      "Combination of adaptive mesh refinement and dealii::MappingQCache not implemented."));
+
+  pde_operator->setup_after_coarsening_and_refinement();
+
+  postprocessor->setup(*pde_operator);
+}
+
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::do_adaptive_refinement(unsigned int const time_step_number)
+{
+  // AMR is only implemented for implicit timestepping.
+  if(trigger_coarsening_and_refinement_now(
+       application->get_parameters().amr_data.trigger_every_n_time_steps, time_step_number))
+  {
+    std::shared_ptr<TimeIntBDF<dim, Number>> bdf_time_integrator =
+      std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+    AssertThrow(bdf_time_integrator.get(),
+                dealii::ExcMessage("Combination of adaptive mesh refinement "
+                                   "and explicit time integration not implemented."));
+
+    mark_cells_coarsening_and_refinement(*grid->triangulation,
+                                         bdf_time_integrator->get_solution_np());
+
+    limit_coarsening_and_refinement(*grid->triangulation, application->get_parameters().amr_data);
+
+    if(any_cells_flagged_for_coarsening_or_refinement(*grid->triangulation))
+    {
+      grid->triangulation->prepare_coarsening_and_refinement();
+
+      time_integrator->prepare_coarsening_and_refinement();
```

Comment:
related to this comment: https://github.com/exadg/exadg/pull/586/files#r1404332107
```C++
if(unsteady)
  time_integrator->prepare_coarsening_and_refinement();
else
  driver_steady->prepare_coarsening_and_refinement();
```

should allow us to use the same function `do_adaptive_refinement()` for both the steady and unsteady case. Of course, we can simply ignore the steady case for now, but comment https://github.com/exadg/exadg/pull/586/files#r1404332107 could already be realized.